### ISSUE TITLE
Installing SU2 on a machine without internet access

### ISSUE THREAD

**Issue by SumanVajjala** [su2code/SU2#955]
Hi,

I am trying to install SU2 from source, on a cluster that has no internet access. Firstly, is it possible to install SU2 in this scenario? Secondly, can you suggest a way to do this without running into dependency issues? I tried to find this information but could not get it.

I will appreciate your help/suggestions regarding this.

Regards
Suman

**Comment by talbring** [comment]
If no internet connection is available, you have to download the appropriate submodules manually and place them in the root folder:

https://github.com/scicompkl/CoDiPack/archive/501dcf0305df147481630f20ce37c2e624fb351f.zip 
https://github.com/scicompkl/MeDiPack/archive/edde14f9ac4026b72b1e130f61c0a78e8652afa5.zip
https://github.com/mesonbuild/meson/archive/0435691e83fb7172e2a9635d2eb32d5521089916.zip
https://github.com/ninja-build/ninja/archive/2d15b04e411229cb902332957281622119025e77.zip 

This information should be printed when no internet connection is detected, unfortunately there is a small bug in the current master branch that prevents that. A fix will be included in the next release. 

Otherwise, you can also just run meson.py once on a computer with internet and then copy it to the other machine.

**Comment by SumanVajjala** [comment]
Hello @talbring. Thank you for the suggestion. After downloading the submodules manually, should I run meson.py or a different script? And when you mean the root folder, do you mean "externals" folder or the system's root folder? If you mean the latter, then I don't have root access. Is this necessary for installation? 
Regards
Suman

**Comment by talbring** [comment]
With root folder I mean the source code root folder of SU2. You can just run meson.py again.

**Comment by SumanVajjala** [comment]
@talbring It worked. Thanks very much. 

Also, how do I enable Intel compilers and Intel MPI for building SU2? Is it via setting CC,CXX,MPICC,MPICXX flags? Where do I specify them?

Regards
Suman

**Comment by talbring** [comment]
Yes you can set them with the usual environment flags, if they are not found automatically

**Comment by SumanVajjala** [comment]
@talbring How do I clean and rebuild the configuration?

Regards
Suman

**Comment by talbring** [comment]
Just delete the configuration folder that you used in the meson call (probably "build" if you followed the documentation).

**Comment by SumanVajjala** [comment]
@talbring I am successful with configuring SU2 using meson.py. However, while installing using ninja I get the following error

[100/846] Compiling C++ object 'externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/ClassicZoneWriterAbstract.cpp.o'.
FAILED: externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/ClassicZoneWriterAbstract.cpp.o 
icpc -Iexternals/tecio/teciompisrc/c3ca80e@@teciompi@sta -Iexternals/tecio/teciompisrc -I../externals/tecio/teciompisrc -Iexternals/tecio/teciompisrc/../ -I../externals/tecio/teciompisrc/../ -Iexternals/tecio/teciompisrc/./ -I../externals/tecio/teciompisrc/./ -I/app/intel/intel2018/impi/2018.3.222/intel64/include -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O3 -fPIC -Wno-unused-parameter -Wno-empty-body -Wno-format-security -Wno-deprecated-declarations -Wno-non-virtual-dtor -fpermissive -DUSEENUM -DTECIOMPI -DTHREED -DMAKEARCHIVE -DNO_ASSERTS -DNO_THIRD_PARTY_LIBS -DTP_PROJECT_USES_BOOST -DBOOST_ALL_NO_LIB -DLINUX -DLINUX64 -MD -MQ 'externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/ClassicZoneWriterAbstract.cpp.o' -MF 'externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/ClassicZoneWriterAbstract.cpp.o.d' -o 'externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/ClassicZoneWriterAbstract.cpp.o' -c ../externals/tecio/teciompisrc/ClassicZoneWriterAbstract.cpp
icpc: command line warning #10006: ignoring unknown option '-Wno-empty-body'
In file included from /usr/local/include/boost/type_traits/is_complete.hpp(15),
                 from /usr/local/include/boost/type_traits/is_convertible.hpp(17),
                 from ../externals/tecio/teciompisrc/../boost/iterator/interoperable.hpp(13),
                 from ../externals/tecio/teciompisrc/../boost/iterator/iterator_facade.hpp(11),
                 from /usr/local/include/boost/range/iterator_range_core.hpp(27),
                 from /usr/local/include/boost/range/iterator_range.hpp(13),
                 from ../externals/tecio/teciompisrc/../boost/assign/list_of.hpp(20),
                 from ../externals/tecio/teciompisrc/../boost/assign.hpp(20),
                 from ../externals/tecio/teciompisrc/fileStuff.h(7),
                 from ../externals/tecio/teciompisrc/ZoneHeaderWriterAbstract.h(3),
                 from ../externals/tecio/teciompisrc/ClassicZoneHeaderWriter.h(2),
                 from ../externals/tecio/teciompisrc/ClassicZoneWriterAbstract.h(2),
                 from ../externals/tecio/teciompisrc/ClassicZoneWriterAbstract.cpp(1):
/usr/local/include/boost/type_traits/is_function.hpp(23): catastrophic error: cannot open source file "boost/type_traits/detail/is_function_cxx_03.hpp"
  #include <boost/type_traits/detail/is_function_cxx_03.hpp>
                                                            ^
Is boost required for installing SU2? I have compiled this using Intel parallel studio 2018.

Regards
Suman

**Comment by talbring** [comment]
Boost is shipped and extracted when meson is run for the first time. Probably it was not extracted succesful somehow. Try to delete the folder externals/tecio/boost and run meson and ninja again.

**Comment by SumanVajjala** [comment]
@talbring It doesn't work. I still get an error with the boost library during TECIO installation. What do you suggest I do?

Regards
Suman

**Comment by talbring** [comment]
There seems to be a strange problem with finding some of the shipped boost headers. For some of them it only finds the system-wide installed boost headers ... actually I have no idea at the moment why this is the case. Can you check whether the file

`externals/tecio/boost/range/iterator_range.hpp` 

exists ? Thats the first occurence where it uses the system installed version ...

**Comment by talbring** [comment]
Otherwise, if you can live without the tecplot output at the moment, use `-Denable-tecio=false` in the meson call to disable it.

**Comment by SumanVajjala** [comment]
@talbring Thanks for the reply. The file exists. I now have a different error.

#error "This Intel <math.h> is for use with only the Intel compilers!"

Regards
Suman

**Comment by SumanVajjala** [comment]
@talbring I have found other issues now (disabled tecio,cgns).

[87/331] Compiling C++ object 'SU2_CFD/src/37e1b17@@SU2_CFD@exe/output_CHeatOutput.cpp.o'.
FAILED: SU2_CFD/src/37e1b17@@SU2_CFD@exe/output_CHeatOutput.cpp.o
icpc -ISU2_CFD/src/37e1b17@@SU2_CFD@exe -ISU2_CFD/src -I../SU2_CFD/src -I../externals/CLI11 -I../externals/metis/include -I../externals/metis/GKlib -I../externals/metis/libmetis -I../externals/parmetis/include -I../externals/parmetis/libparmetis -ICommon/src/./ -I../Common/src/./ -I/app/intel/intel2018/impi/2018.3.222/intel64/include -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O3 -fPIC -Wno-unused-parameter -Wno-empty-body -Wno-format-security -Wno-deprecated-declarations -Wno-non-virtual-dtor -DNDEBUG -DHAVE_MPI -DHAVE_METIS -DHAVE_PARMETIS -MD -MQ 'SU2_CFD/src/37e1b17@@SU2_CFD@exe/output_CHeatOutput.cpp.o' -MF 'SU2_CFD/src/37e1b17@@SU2_CFD@exe/output_CHeatOutput.cpp.o.d' -o 'SU2_CFD/src/37e1b17@@SU2_CFD@exe/output_CHeatOutput.cpp.o' -c ../SU2_CFD/src/output/CHeatOutput.cpp
icpc: command line warning #10006: ignoring unknown option '-Wno-empty-body'
../SU2_CFD/src/output/../../../Common/include/geometry/../toolboxes/allocation_toolbox.hpp(79): **error: the global scope has no "aligned_alloc"
    ptr = ::aligned_alloc(alignment, size);**
            ^
          detected during:
            instantiation of "void container_helpers::AccessorImpl<Index_t, Scalar_t, Store, AlignSize, 0UL, 0UL>::m_allocate(size_t={unsigned long}, Index_t, Index_t) [with Index_t=unsigned long, Scalar_t=su2double={double}, Store=StorageType::RowMajor, AlignSize=64UL]" at line 411 of "../SU2_CFD/src/output/../../../Common/include/geometry/../toolboxes/C2DContainer.hpp"
            instantiation of "size_t={unsigned long} C2DContainer<Index_t, Scalar_t, Store, AlignSize, StaticRows, StaticCols>::m_resize(Index_t, Index_t) [with Index_t=unsigned long, Scalar_t=su2double={double}, Store=StorageType::RowMajor, AlignSize=64UL, StaticRows=0UL, StaticCols=0UL]" at line 445 of "../SU2_CFD/src/output/../../../Common/include/geometry/../toolboxes/C2DContainer.hpp"
            instantiation of "C2DContainer<Index_t, Scalar_t, Store, AlignSize, StaticRows, StaticCols> &C2DContainer<Index_t, Scalar_t, Store, AlignSize, StaticRows, StaticCols>::operator=(const C2DContainer<Index_t, Scalar_t, Store, AlignSize, StaticRows, StaticCols> &) [with Index_t=unsigned long, Scalar_t=su2double={double}, Store=StorageType::RowMajor, AlignSize=64UL, StaticRows=0UL, StaticCols=0UL]" at line 63 of
                      "../SU2_CFD/src/output/../../include/solvers/../../../Common/include/geometry/elements/CGaussVariable.hpp"

compilation aborted for ../SU2_CFD/src/output/CHeatOutput.cpp (code 2)

Could the issue be because of C++-11 features not being supported by the compiler I am using? I am using Intel Parallel Studio 2017.

Regards
Suman

**Comment by pcarruscag** [comment]
Can't you use gcc? You will get worse performance with any version of the intel compilers prior to 2019 than with even gcc 5.4...
If this is because you need intel MPI you can do:
export CC=gcc
export CXX=g++
export I_MPI_CC=gcc
export I_MPI_CXX=g++
And then follow the normal installation procedure (meson ninja). These problems you are running into with boost also have a "half baked c++11 compiler support" smell to them, namely the 03 suffix to that file.

**Comment by SumanVajjala** [comment]
@pcarruscag @talbring  I am compiling with gcc 5.5.0 and mpich3.2.1. I am getting this error

[107/846] Compiling C++ object 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysVector.cpp.o'.
FAILED: Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysVector.cpp.o 
c++ -ICommon/src/9e112f9@@SU2Common@sta -ICommon/src -I../Common/src -I../externals/CLI11 -I../externals/cgns/adf -Iexternals/cgns/./ -I../externals/cgns/./ -I../externals/metis/include -I../externals/metis/GKlib -I../externals/metis/libmetis -I../externals/parmetis/include -I../externals/parmetis/libparmetis -Iexternals/tecio/teciompisrc/../ -I../externals/tecio/teciompisrc/../ -Iexternals/tecio/teciompisrc/./ -I../externals/tecio/teciompisrc/./ -I/app/mpich-3.2.1/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O3 -fPIC -Wl,-rpath -Wl,/app/mpich-3.2.1/lib -Wl,--enable-new-dtags -Wl,-rpath -Wl,/app/mpich-3.2.1/lib -Wl,--enable-new-dtags -Wno-unused-parameter -Wno-empty-body -Wno-format-security -Wno-deprecated-declarations -Wno-non-virtual-dtor -DHAVE_CGNS -DNDEBUG -DHAVE_MPI -DHAVE_METIS -DHAVE_PARMETIS -DHAVE_TECIO -DTECIOMPI -MD -MQ 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysVector.cpp.o' -MF 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysVector.cpp.o.d' -o 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysVector.cpp.o' -c ../Common/src/linear_algebra/CSysVector.cpp
In file included from ../Common/src/linear_algebra/CSysVector.cpp:31:0:
../Common/src/linear_algebra/../../include/toolboxes/allocation_toolbox.hpp: In function ‘T* MemoryAllocation::aligned_alloc(size_t, size_t)’:
../Common/src/linear_algebra/../../include/toolboxes/allocation_toolbox.hpp:79:9: error: ‘::aligned_alloc’ has not been declared
   ptr = ::aligned_alloc(alignment, size);
         ^
../Common/src/linear_algebra/../../include/toolboxes/allocation_toolbox.hpp:79:9: note: suggested alternative:
../Common/src/linear_algebra/../../include/toolboxes/allocation_toolbox.hpp:61:11: note:   ‘MemoryAllocation::aligned_alloc’
 inline T* aligned_alloc(size_t alignment, size_t size) noexcept
           ^
[111/846] Compiling C++ object 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysSolve.cpp.o'.
FAILED: Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysSolve.cpp.o 
c++ -ICommon/src/9e112f9@@SU2Common@sta -ICommon/src -I../Common/src -I../externals/CLI11 -I../externals/cgns/adf -Iexternals/cgns/./ -I../externals/cgns/./ -I../externals/metis/include -I../externals/metis/GKlib -I../externals/metis/libmetis -I../externals/parmetis/include -I../externals/parmetis/libparmetis -Iexternals/tecio/teciompisrc/../ -I../externals/tecio/teciompisrc/../ -Iexternals/tecio/teciompisrc/./ -I../externals/tecio/teciompisrc/./ -I/app/mpich-3.2.1/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O3 -fPIC -Wl,-rpath -Wl,/app/mpich-3.2.1/lib -Wl,--enable-new-dtags -Wl,-rpath -Wl,/app/mpich-3.2.1/lib -Wl,--enable-new-dtags -Wno-unused-parameter -Wno-empty-body -Wno-format-security -Wno-deprecated-declarations -Wno-non-virtual-dtor -DHAVE_CGNS -DNDEBUG -DHAVE_MPI -DHAVE_METIS -DHAVE_PARMETIS -DHAVE_TECIO -DTECIOMPI -MD -MQ 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysSolve.cpp.o' -MF 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysSolve.cpp.o.d' -o 'Common/src/9e112f9@@SU2Common@sta/linear_algebra_CSysSolve.cpp.o' -c ../Common/src/linear_algebra/CSysSolve.cpp
In file included from ../Common/src/linear_algebra/../../include/geometry/../toolboxes/C2DContainer.hpp:30:0,
                 from ../Common/src/linear_algebra/../../include/geometry/../toolboxes/graph_toolbox.hpp:30,
                 from ../Common/src/linear_algebra/../../include/geometry/CGeometry.hpp:61,
                 from ../Common/src/linear_algebra/CSysSolve.cpp:33:
../Common/src/linear_algebra/../../include/geometry/../toolboxes/allocation_toolbox.hpp: In function ‘T* MemoryAllocation::aligned_alloc(size_t, size_t)’:
../Common/src/linear_algebra/../../include/geometry/../toolboxes/allocation_toolbox.hpp:79:9: error: ‘::aligned_alloc’ has not been declared
   ptr = ::aligned_alloc(alignment, size);
         ^
../Common/src/linear_algebra/../../include/geometry/../toolboxes/allocation_toolbox.hpp:79:9: note: suggested alternative:
../Common/src/linear_algebra/../../include/geometry/../toolboxes/allocation_toolbox.hpp:61:11: note:   ‘MemoryAllocation::aligned_alloc’
 inline T* aligned_alloc(size_t alignment, size_t size) noexcept
           ^
[124/846] Compiling C++ object 'externals/tecio/teciompisrc/c3ca80e@@teciompi@sta/importSzPltFile.cpp.o'.
ninja: build stopped: subcommand failed.

Looks like the aligned_alloc is causing this issue. Please suggest what I need to do to get past this error.

Regards
Suman

**Comment by pcarruscag** [comment]
I have no idea, that function is not a problem even with gcc 4.7.3, and I also use a similar version of mpich on some machines.
a) Clean the code directory, delete the whole thing and re-download the code just to be sure.
b) What OS are you on?
c) Run mpirun --version, mpicxx --version post the output here.
d) Run the configuration with meson, post the output here.

**Comment by SumanVajjala** [comment]
@pcarruscag 

a) I have done that. I get the same error.
b) Linux is the OS
c) The output of mpirun --version is
HYDRA build details:
    Version:                                 3.2.1
    Release Date:                            Fri Nov 10 20:21:01 CST 2017
    CC:                              gcc    
    CXX:                             g++    
    F77:                             gfortran   
    F90:                             gfortran   
    Configure options:                       '--disable-option-checking' '--prefix=/app/mpich-3.2.1' '--cache-file=/dev/null' '--srcdir=.' 'CC=gcc' 'CFLAGS= -O2' 'LDFLAGS=' 'LIBS=-lpthread ' 'CPPFLAGS= -I/app/setups/mpich-3.2.1/src/mpl/include -I/app/setups/mpich-3.2.1/src/mpl/include -I/app/setups/mpich-3.2.1/src/openpa/src -I/app/setups/mpich-3.2.1/src/openpa/src -D_REENTRANT -I/app/setups/mpich-3.2.1/src/mpi/romio/include' 'MPLLIBNAME=mpl'
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Checkpointing libraries available:       
    Demux engines available:                 poll select

The output of mpicxx --version is

g++ (GCC) 5.5.0
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

d) The configuration of meson:

./meson.py build -Denable-autodiff=true -Denable-directdiff=true -Denable-tecio=false -Dwith-mpi=disabled -Denable-cgns=false --prefix=/home/vksuman/SU2_7_0_3_gcc5_5_0_mpich3_2_1
ninja executable not found. Building ...
bootstrapping ninja...
warning: A compatible version of re2c (>= 0.11.3) was not found; changes to src/*.in.cc will not affect your build.
wrote build.ninja.
bootstrap complete.  rebuilding...
[29/29] LINK ninja
The Meson build system
Version: 0.52.999
Source dir: /home/vksuman/SU2-7.0.3
Build dir: /home/vksuman/SU2-7.0.3/build
Build type: native build
Project name: SU2
Project version: 7.0.3 "Blackbird"
C compiler for the host machine: gcc (gcc 5.5.0 "gcc (GCC) 5.5.0")
C linker for the host machine: GNU ld.bfd 2.20.51.0.2-5
C++ compiler for the host machine: g++ (gcc 5.5.0 "g++ (GCC) 5.5.0")
C++ linker for the host machine: GNU ld.bfd 2.20.51.0.2-5
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/app/python-3.6.3/bin/python3.6)
Dependency mpi skipped: feature with-mpi disabled
Dependency mpi skipped: feature with-mpi disabled
Message: -------------------------------------------------------------------------
         |    ___ _   _ ___                                                      |
         |   / __| | | |_  )   Release 7.0.3 "Blackbird"                         |
         |   \__ \ |_| |/ /                                                      |
         |   |___/\___//___|   Meson Configuration Summary                       |
         |                                                                       |
         -------------------------------------------------------------------------
         
         Option          Value
         ---------------------
         TecIO:          false
         CGNS:           false
         AD (reverse):   true
         AD (forward):   true
         Python Wrapper: false
         Intel-MKL:      false
         OpenBlas:       false
         PaStiX:         false

         Please be sure to add the $SU2_HOME and $SU2_RUN environment variables,
         and update your $PATH (and $PYTHONPATH if applicable) with $SU2_RUN

         Based on the input to this configuration, add these lines to your .bashrc file:

         export SU2_RUN=/home/vksuman/SU2_7_0_3_gcc5_5_0_mpich3_2_1/bin
         export SU2_HOME=/home/vksuman/SU2-7.0.3
         export PATH=$PATH:$SU2_RUN
         export PYTHONPATH=$PYTHONPATH:$SU2_RUN

         Use './ninja -C build install' to compile and install SU2

Build targets in project: 12
Found ninja-1.9.0.git at /home/vksuman/SU2-7.0.3/ninja

**Comment by SumanVajjala** [comment]
@pcarruscag I am installing this on a cluster with no internet access. So, I followed the advice from @talbring and I downloaded the codipack, medipack, meson and ninja on my local computer (with internet access). I then replaced the directories in the externals folder in the cluster with the downloaded directories. The local computer runs Ubuntu and used python3 to download the libraries. I hope this is okay.

Regards
Suman

**Comment by pcarruscag** [comment]
I see... well maybe there is something wrong with the compiler installation, or how the default include directories are setup, gcc 5.5 has no problems finding aligned_alloc : https://gcc.godbolt.org/z/GGYrft
If that cluster is managed by someone, your best option is to get support from them.

**Comment by SumanVajjala** [comment]
@economon I have personally installed SU2 software (from versions 3 to 6) in our cluster without problems. I think the issues could be because of lack of C11 features. I tried to run a sample code with aligned_alloc and I had problems with both Intel and GCC compilers. I'll test this again. I think I'll have to install gcc myself.

Regards
Suman

**Comment by erangit** [comment]
@SumanVajjala,
you may want to try to use the older preconfig/mame/make install procedure.
Just do the preconfig process (with -u parameter)  on a machine that is
connected to the internet so codi/medi versions will be updated
Next step: search the preconfig.py file  for the codi/medi version (look
for "sha_version_codi=" and "sha_version_medi=").
Write down the two strings somewhere.
Then transfer the whole install folder to the disconnected cluster.
Once on the cluster, go to folder external/codi and touch a file by the
name of the string you have written down.
Repeat the same with medi (external/medi).
Now do on the cluster preconfig.py/make/make install (without -u). The
codi/medi versions will match now and the installation should work.
The problem with this procedure is that it is not supported anymore. It
still worked for me a short time ago with the develop branch (but failed on
the master branch). Meson/ninja installation is significantly more
efficient and most of the developers here ain't familiar with life without
an internet connection ...

Last warning: the above-listed messages seem to point over a compiler
issue. If indeed this is the case then the bypass that I suggested is
irrelevant.

**Comment by pcarruscag** [comment]
@SumanVajjala gcc 5+ have full c++ 11 support (even more than that actually). As a last resort, if you cannot figure out what is going on with the compilers (simpler guess is that there are other versions installed and they are getting mixed up?), you can try replacing the file "allocation_toolbox.hpp" by this:
[allocation_toolbox_PATCH.txt](https://github.com/su2code/SU2/files/4550786/allocation_toolbox_PATCH.txt)

**Comment by SumanVajjala** [comment]
> @SumanVajjala gcc 5+ have full c++ 11 support (even more than that actually). As a last resort, if you cannot figure out what is going on with the compilers (simpler guess is that there are other versions installed and they are getting mixed up?), you can try replacing the file "allocation_toolbox.hpp" by this:
> [allocation_toolbox_PATCH.txt](https://github.com/su2code/SU2/files/4550786/allocation_toolbox_PATCH.txt)

@pcarruscag Thank you very much. This works. I have installed a sequential version and a parallel version with tecio,codipack and medipack enabled. I will try the same with Intel compilers and see if it works.

Regards
Suman

**Comment by GomerOfDoom** [comment]
I realize this is closed, but I encountered the same issue where aligned_alloc is not declared in the global namespace. I'm writing this post just in case anyone else runs into this problem on a cluster computer where they may not have the ability to change or install different compilers or c/c++ libraries.

The problem stems from the version of cstdlib (or stdlib.h) and the c++ (or c) standard that is specified.

alligned_alloc was not introduced in c++11, it was introduced in c11, and didn't make it into the c++ standard until c++17.

My understanding is that basically, if stdlib.h is used, and the c (not c++) standard is set to 2011 or greater, aligned_alloc should appear in the global namespace. If cstdlib is used and the c++ standard is set to 2017 or greater, then aligned_alloc will appear in the std namespace. Also, I believe if the c++ standard is set to 2017 or greater, and stdlib.h is used, aligned_alloc will also appear in the global namespace.

However, there were several gcc compiler releases (including up to 7.3) that did not have aligned_alloc as part of the std:: namespace. Apparently, this bug was fixed in gcc 7.4 and up.

I was also able to work around this problem by using the allocation_toolbox_PATCH.txt that @pcarruscag provided, above, as well.

-Paul

**Comment by pcarruscag** [comment]
Hi Paul, indeed when I introduced aligned allocation I took for granted that it was a c++11 feature.
Did you come across any programmatic way to detect whether aligned_alloc is available? We could have a graceful fallback to normal new/delete, no part of the code strictly relies on alignment (yet).

**Comment by GomerOfDoom** [comment]
Realized I never followed up. I did not find a programmatic way to detect this. I think the best solution is to either have people use your work-around file if they are stuck with an old compiler.

When aligned_alloc becomes a required part of SU2, we may need to include a minimum compiler version requirement for the code. IMO, it is too much to continually write code that supports ancient compiler versions.

**Comment by erangit** [comment]
HI 
This issue was closed, but while the PATCH from April 29, 2020, was effective on disconnected machines with compilers (GCC) that are not fully updated, it stopped working on the develop branch (a few months ago it did the work). It should be said that on updated GCC (Ubuntu), with a network connection, the patch is not needed.  Any idea how to update the PATCH or a different solution?
Thnx, Eran

**Comment by pcarruscag** [comment]
Please attach the compiler warning you are getting.

**Comment by pcarruscag** [comment]
errors, compiler errors

**Comment by erangit** [comment]
[
[errors_SU2_71.pdf](https://github.com/su2code/SU2/files/6914931/errors_SU2_71.pdf)
Hi 
Attached is a printout of errors (part of it, it keeps repeating). It seems to me that it is not related to the problem that was resolved by the patch, mentioned above: it happens before during the compilation, It happens also both with and without the patch. This one is from the master version, downloaded a few days ago. Had the same errors from develop branch, a few days earlier. Needless to say that this issue arises only with GCC that is not up-to-date. On my web-connected laptop, everything compiles and links fine. Sorry for the delay in response and the format of the printout. It takes time to get data out of the disconnected network.

**Comment by pcarruscag** [comment]
That seems to be a hard one, what is the GCC version?

**Comment by erangit** [comment]
4.8.5 
Yep, that is pretty old (got 7.5.0 on my Laptop). However, till a month or two ago it did the work. We are also compiling other updated codes (like CharLES) with it without problem

**Comment by pcarruscag** [comment]
Have you tried adding this "-fabi-version=6" to the CXXFLAGS? I was able to reproduce the errors you shared and this flag seems to fix it.

**Comment by erangit** [comment]
Halleluja, you did it again. Compiled and linked properly !!!

**Comment by erangit** [comment]
I think we can close this issue again, till next time, as the issue is resolved. I shall close it soon, unless, Pedro, you prefer to leave it open for a while, so that it is visible to other people with a similar problem

**Comment by pcarruscag** [comment]
I want to try making meson more offline-friendly.

**Comment by Aslihan35** [comment]
Hi,

I try to install SU2-7.4.0 with mpi and cgns features to Ubuntu 22.04.3 LTS. I run "./meson.py build <arguments>" successfully,
however, the installation stops at the following step while running "./ninja -C build essential".

"[461/1361] Compiling C++ object extern...ibteciompi.a.p/ZoneWriterFactory.cpp.o"

I also tried to install it without mpi, cgns and tecio, but unfortunately it stops at the following step.

"[6/842] Compiling C++ object Common/src/libSU2Common.a.p/linear_algebra_CSysSolve.cpp.o"

I used the patch provided on this page, but unfortunately, it didn't resolve the issue. I use gcc 9.5.0, g++ 9.5.0, mpirun 4.1.2, and python 3.10.12 for installation. 

Any help would be appreciated.
Thanks,
